libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,256 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * @file Inliner.cpp
+ * Inlines small code snippets by replacing JUMP with a copy of the code jumped to.
+ */
+
+#include <libevmasm/Inliner.h>
+
+#include <libevmasm/AssemblyItem.h>
+#include <libevmasm/GasMeter.h>
+#include <libevmasm/KnownState.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/numeric/accumulate.hpp>
+#include <range/v3/view/drop_last.hpp>
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/slice.hpp>
+#include <range/v3/view/transform.hpp>
+
+#include <optional>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::evmasm;
+
+
+namespace
+{
+/// @returns an estimation of the runtime gas cost of the AsssemblyItems in @a _itemRange.
+template<typename RangeType>
+u256 executionCost(RangeType&& _itemRange, langutil::EVMVersion _evmVersion)
+{
+	GasMeter gasMeter{std::make_shared<KnownState>(), _evmVersion};
+	auto gasConsumption = ranges::accumulate(std::forward<RangeType>(_itemRange) | ranges::views::transform(
+		[&gasMeter](auto const& _item) { return gasMeter.estimateMax(_item, false); }
+	), GasMeter::GasConsumption());
+	if (gasConsumption.isInfinite)
+		return numeric_limits<u256>::max();
+	else
+		return gasConsumption.value;
+}
+/// @returns an estimation of the code size in bytes needed for the AssemblyItems in @a _itemRange.
+template<typename RangeType>
+uint64_t codeSize(RangeType&& _itemRange)
+{
+	return ranges::accumulate(std::forward<RangeType>(_itemRange) | ranges::views::transform(
+		[](auto const& _item) { return _item.bytesRequired(2); }
+	), 0u);
+}
+/// @returns the tag id, if @a _item is a PushTag or Tag into the current subassembly, nullopt otherwise.
+optional<size_t> getLocalTag(AssemblyItem const& _item)
+{
+	if (_item.type() != PushTag && _item.type() != Tag)
+		return nullopt;
+	auto [subId, tag] = _item.splitForeignPushTag();
+	if (subId != numeric_limits<size_t>::max())
+		return nullopt;
+	return tag;
+}
+}
+
+bool Inliner::isInlineCandidate(size_t _tag, ranges::span<AssemblyItem const> _items) const
+{
+	assertThrow(_items.size() > 0, OptimizerException, "");
+
+	// Only consider blocks that end in a JUMP for now. This can e.g. be extended to include transaction terminating
+	// instructions as well in the future.
+	if (_items.back() != Instruction::JUMP)
```

Comment:
```suggestion
	if (_items.back() != Instruction::JUMP && _items.back().getJumpType() == AssemblyItem::JumpType::OutOfFunction)
```
Can't we check the jump type early?

Reply:
Only because we only inline blocks ending in ``OutOfFunction`` right now, does not mean that we could not inline other blocks as well, so I'd say we collect all candidates right now and only choose to inline specific jump type combinations later. We can at any time extend ``Inliner::shouldInline`` to include more combinations. And it's not like this will be performance critical - all of this is extremely fast anyways.

Reply:
Why not remove the check on the jumptype when we are extending the inliner?

Reply:
But I'm fine with both options.

Reply:
Well, why build things in such a way that you'll need to change two places later instead of one :-)? It's not like the check in ``shouldInline`` would vanish, because the other half of the jump type we only know there anyways. So I'd say the split makes sense as it is. If it's a bother, I can also remove this preselection entirely and check everything in ``shouldInline`` instead...